### ISSUE TITLE
[proposal] umoci-revert?

### ISSUE THREAD

**Issue by cyphar** [openSUSE/umoci#172]
At the moment if someone wants to create several variants of a layer on top of some base image, they have to call `umoci unpack` many times (which is wasteful and slow because of how big most images are). The key question is can we effectively use `mtree` to revert changes to the `rootfs` (extracting files if necessary from the blob layers) so that we can return to a pristine `rootfs`?

* [ ] Refactor `restoreMetadata` and the other related code to be usable by an `mtreeRevert`.
* [ ] Create an index of extracted files during `unpack` to allow us to know what layer blob (and offset) we need to read to extract the image.
* [ ] Lots and lots of testing.

/cc @vbatts since he said he was planning on doing this in go-mtree. I think that because we have `mtree.Compare` we can do it inside `umoci` but maybe he has some comments on this.

**Comment by vbatts** [comment]
I think this would play well with the concept of understanding the now v1
OCI image layer. Applying a  changeset to the manifest, to arrive at an
expected representation. Right? So to revert, one needs to apply a
changeset to the manifest structure, then compare.

**Comment by hallyn** [comment]
Realistically I assume everyone just uses overlay or a snapshotting fs for this.

**Comment by cyphar** [comment]
Yeah, I think #574 and possibly #325 are better approaches for this problem...
